### PR TITLE
adapter: refactor max connection handling

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR refactors our max connection handling in a few ways:

1. It pulls the logic out of the `sql` crate and moves it into the `pgwire_common` crate. AFAICT this logic lived in the `sql` crate so it could be updated via `SystemVars`.
2. Refactors `SystemVars` to contain a generic set of callbacks so anyone can register to be notified about changes to a system variable, not just connection handling.

It also fixes a bug that has occurred in production. Previously if the max connection limit dropped below the number of current connections (e.g. reducing the LaunchDarkly flag) we could panic.

### Motivation

Fixes a known [panic](https://materializeinc.sentry.io/auth/login/materializeinc/?next=%2Fissues%2F5979946058%2Fevents%2Fff772de8c8b34c3bb11b240b93745222%2F%3Fproject%3D6780145) and refactors the code.

### Tips for reviewer

I would start by reviewing the changes in `sql/src/session/vars.rs` to see how SystemVars changed, then look at the changes in `pgwire-common/src/conn.rs`.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/pgwire-common/src/conn.rs [hunk 1]
```diff
@@ -117,3 +119,147 @@ where
...
+    /// Current number of connections.
+    current: u64,
+    /// Total number of connections allowed.
+    limit: u64,
+    /// Number of connections we'll reserve for super users.
+    superuser_reserved: u64,
+}
+
+impl ConnectionCounterInner {
+    fn new(limit: u64, superuser_reserved: u64) -> Self {
```
Comment (jkosh44) [inline src/pgwire-common/src/conn.rs:209]: We might benefit from an assert like `assert!(superuser_reserved < limit)`

File: src/pgwire-common/src/conn.rs [hunk 2]
```diff
@@ -117,3 +119,147 @@ where
...
+    }
+}
+
+#[derive(Debug)]
+pub struct ConnectionCounterInner {
+    /// Current number of connections.
+    current: u64,
+    /// Total number of connections allowed.
+    limit: u64,
+    /// Number of connections we'll reserve for super users.
```
Comment (jkosh44) [inline src/pgwire-common/src/conn.rs:204]: It took me awhile to figure out that this is the number of connections in `limit`. Maybe something like
```suggestion
    /// Number of connections in `limit` we'll reserve for super users.
```
would make it clearer.

File: src/adapter/src/coord.rs [hunk 3]
```diff
@@ -3933,6 +3932,28 @@ pub fn serve(
...
+        // Register a callback so whenever the MAX_CONNECTIONS or SUPERUSER_RESERVED_CONNECTIONS
+        // system variables change, we update our connection limits.
+        let connection_limit_callback: Arc<dyn Fn(&SystemVars) + Send + Sync> =
+            Arc::new(move |system_vars: &SystemVars| {
+                let limit: u64 = system_vars.max_connections().cast_into();
+                let superuser_reserved: u64 =
+                    system_vars.superuser_reserved_connections().cast_into();
+
+                // If superuser_reserved > max_connections, prefer max_connections.
+                let superuser_reserved = std::cmp::min(limit, superuser_reserved);
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:3944]: If `superuser_reserved > max_connections` is true, then that means only super users could connect to the database right? i.e., all connections are reserved for superusers and normal users will be locked out. That feels like a bad enough situation that we should add an error log here.

**Review by jkosh44 — APPROVED** [review]
LGTM, had a couple of small questions and nits. It also might be good to add a test where we lower the connection limit with a bunch of connections open to repro the panic.

File: src/pgwire-common/src/conn.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/pgwire-common/src/conn.rs:209]: makes sense, added!

File: src/pgwire-common/src/conn.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/pgwire-common/src/conn.rs]: Updated!

File: src/adapter/src/coord.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/adapter/src/coord.rs]: Great call, added a soft assert here!

**Comment by ParkMyCar** [comment]
Added a test that previously repro'd the panic!

**Pull request merged**